src/ASM/ASMs2Dmx.C

```diff
@@ -1026,3 +1027,112 @@ void ASMs2Dmx::getBoundaryNodes (int lIndex, IntVec& nodes, int basis,
     for (size_t b = 1; b <= this->getNoBasis(); ++b)
       this->ASMs2D::getBoundaryNodes(lIndex, nodes, b, thick, local);
 }
+
+
+bool ASMs2Dmx::assembleL2matrices (SparseMatrix& A, StdVector& B,
+                                   const IntegrandBase& integrand,
+                                   bool continuous) const
+{
+  const size_t nnod = this->getNoNodes(1);
+
+  const int p1 = surf->order_u();
+  const int p2 = surf->order_v();
+  const int n1 = surf->numCoefs_u();
+  const int n2 = surf->numCoefs_v();
+  const int nel1 = n1 - p1 + 1;
+  const int nel2 = n2 - p2 + 1;
+
+  // Get Gaussian quadrature point coordinates (and weights if continuous)
+  const int ng1 = continuous ? nGauss : p1 - 1;
+  const int ng2 = continuous ? nGauss : p2 - 1;
+  const double* xg = GaussQuadrature::getCoord(ng1);
+  const double* yg = GaussQuadrature::getCoord(ng2);
+  const double* wg = continuous ? GaussQuadrature::getWeight(nGauss) : nullptr;
+  if (!xg || !yg) return false;
+  if (continuous && !wg) return false;
+
+  // Compute parameter values of the Gauss points over the whole patch
+  Matrix gp;
+  std::array<RealArray,2> gpar;
+  gpar[0] = this->getGaussPointParameters(gp,0,ng1,xg);
+  gpar[1] = this->getGaussPointParameters(gp,1,ng2,yg);
+
+  // Evaluate basis functions at all integration points
+  std::vector<Go::BasisPtsSf>    spl0, spl01;
+  std::vector<Go::BasisDerivsSf> spl1;
+  if (continuous)
+    surf->computeBasisGrid(gpar[0],gpar[1],spl1);
+  else
+    surf->computeBasisGrid(gpar[0],gpar[1],spl0);
+
+  m_basis[0]->computeBasisGrid(gpar[0],gpar[1],spl01);
+
+  // Evaluate the secondary solution at all integration points
+  Matrix sField;
+  if (!this->evalSolution(sField,integrand,gpar.data()))
+  {
+    std::cerr <<" *** ASMs2D::assembleL2matrices: Failed for patch "<< idx+1
+	      <<" nPoints="<< gpar[0].size()*gpar[1].size() << std::endl;
+    return false;
+  }
+
+  double dA = 1.0;
+  Vector phi(p1*p2), phi2;
+  Matrix dNdu, Xnod, J;
+
+
+  // === Assembly loop over all elements in the patch ==========================
+
+  int iel = 0;
+  for (int i2 = 0; i2 < nel2; i2++)
+    for (int i1 = 0; i1 < nel1; i1++, iel++)
+    {
+      if (MLGE[iel] < 1) continue; // zero-area element
+
+      if (continuous)
+      {
+	// Set up control point (nodal) coordinates for current element
+	if (!this->getElementCoordinates(Xnod,1+iel))
+	  return false;
+	else if ((dA = 0.25*this->getParametricArea(1+iel)) < 0.0)
+	  return false; // topology error (probably logic error)
+      }
+
+      // --- Integration loop over all Gauss points in each direction ----------
+
+      int ip = (i2*ng1*nel1 + i1)*ng2;
+      for (int j = 0; j < ng2; j++, ip += ng1*(nel1-1))
+	for (int i = 0; i < ng1; i++, ip++)
+	{
+	  if (continuous)
+	    SplineUtils::extractBasis(spl1[ip],phi,dNdu);
+	  else
+	    phi = spl0[ip].basisValues;
+
```

Comment:
`phi` is not used, so this statement can be skipped, and you can use phi2 as dummy in the extractBasis call such that you only need the `phi2` variable here.